tailcfg/tailcfg.go

```diff
@@ -1596,3 +1596,30 @@ type SSHAction struct {
 	// value.
 	HoldAndDelegate string `json:"holdAndDelegate,omitempty"`
 }
+
+// OverTLSPublicKeyResponse is the JSON response to /key?v=<n>
+// over HTTPS (regular TLS) to the Tailscale control plane server,
+// where the 'v' argument is the client's current capability version
+// (previously known as the "MapRequest version").
+//
+// The "OverTLS" prefix is to loudly declare that this exchange
+// doesn't happen over Noise and can be intercepted/MITM'ed by
+// enterprise/corp proxies where the orgnanization can put TLS roots
+// on devices.
+type OverTLSPublicKeyResponse struct {
+	// LegacyPublic specifies the control plane server's original
+	// NaCl crypto_box machine key.
+	// It will be zero for sufficiently new clients, based on their
+	// advertised "v" parameter (the CurrentMapRequestVersion).
+	// In that case, only the newer Noise-based transport may be used
+	// using the PublicKey field.
+	LegacyPublicKey key.MachinePublic `json:"legacyPublicKey"`
+
+	// PublicKey specifies the server's public key for the
+	// Noise-based control plane protocol.
+	PublicKey key.MachinePublic `json:"publicKey"`
+
+	// Expiry, if non-nil, is the time at which the public key(s)
+	// expire.
+	Expiry *time.Time `json:"expiry,omitempty"`
```

Comment:
Expiry is an example of something that shouldn't be distributed in plain text. (E.g. when a TLS cert includes an expiry, it's covered by the signature of the certificate.)

Reply:
Removed for now but I still think we want something here like this, perhaps named differently.

Currently the client caches the public key forever and never re-checks. It's TOFU for life of the process, which means we can never rotate the nacl public key. I want a field like this to say that the client should at least re-query the public key periodically, but we can deal with that later.

Reply:
I agree we want an expiry. But the expiry needs proof associated that the server key agrees with it. The easy way for us to do that is to return it in a /meta request inside the noise channel.